### PR TITLE
SALTO-5758 page and space CVs for Confluence

### PR DESCRIPTION
**Pull request by omrifarkash**
Added two change validator to Confluence:
1. Uniqueness of page title under space
2. Uniqueness of space key

---

---
_Release Notes_: 
_Confluence Adapter_:
- CV for uniqueness of page title under space
- CV for uniqueness of space key

---
_User Notifications_: 
_Confluence Adapter_:
- CV for uniqueness of page title under space
- CV for uniqueness of space key

### CODE REVIEW

File: packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    log.warn('elementSource is undefined, skipping uniquePageTitleUnderSpaceValidator')
+    return []
+  }
+  const pageChanges = changes
+    .filter(isAdditionOrModificationChange)
+    .map(getChangeData)
+    .filter(isInstanceElement)
+    .filter(instance => instance.elemID.typeName === PAGE_TYPE_NAME)
+
+  const spaceNameIdToPagesInstances = await awu(await elementSource.getAll()).reduce<Record<string, InstanceElement[]>>(
```
Comment (Nurdok) [inline packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:42]: Grammar nit: spaceNameIdToPageInstances ("Page" should be singular)

File: packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    },
+    {},
+  )
+  const spaceNameToTitlePagesIndex = _.mapValues(spaceNameIdToPagesInstances, pages =>
+    pages.reduce<Record<string, InstanceElement[]>>((record, page) => {
+      const { title } = page.value
+      if (record[title]) {
+        record[title].push(page)
+      } else {
+        record[title] = [page]
```
Comment (Nurdok) [inline packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:62]: Ditto:

```ts
(record[title] ??= []).push(page)
```

File: packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          record[spaceElemId].push(elem)
+        } else {
+          record[spaceElemId] = [elem]
+        }
+      }
+      return record
+    },
+    {},
+  )
+  const spaceNameToTitlePagesIndex = _.mapValues(spaceNameIdToPagesInstances, pages =>
```
Comment (Nurdok) [inline packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:56]: There's `TitlePages` in the name, but shouldn't it be `PageTitle`? Maybe explicitly writing out the type will make this clearer (took me a few minutes to figure out).

File: packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    )
+    if (pageWithTheSameTitle === undefined) {
+      return []
+    }
+    return [
+      {
+        elemID: pageInstFromChange.elemID,
+        severity: 'Error',
+        message: 'Page title must be unique under space',
+        detailedMessage: `Page title: ${pageInstFromChange.value.title} is already in use by page: ${pageWithTheSameTitle.elemID.getFullName()}`,
```
Comment (Nurdok) [inline packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:80]: Maybe surround the page title with quotes?

File: packages/confluence-adapter/src/change_validators/space_key_uniqueness.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+const log = logger(module)
+const { awu } = collections.asynciterable
+
+export const uniqueSpaceKeyValidator: ChangeValidator = async (changes, elementSource) => {
+  if (elementSource === undefined) {
+    log.warn('elementSource is undefined, skipping uniqueSpaceKeyValidator')
+    return []
+  }
+  const spaceChangesInstances = changes
```
Comment (Nurdok) [inline packages/confluence-adapter/src/change_validators/space_key_uniqueness.ts:35]: Grammar nit: should be `spaceChangeInstances` (change is singular)

**Review by Nurdok — APPROVED** [review]

File: packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  const spaceNameIdToPagesInstances = await awu(await elementSource.getAll()).reduce<Record<string, InstanceElement[]>>(
+    (record, elem) => {
+      if (isInstanceElement(elem) && elem.elemID.typeName === PAGE_TYPE_NAME) {
+        const spaceElemId = elem.value.spaceId.elemID.getFullName()
+        if (record[spaceElemId]) {
+          record[spaceElemId].push(elem)
+        } else {
+          record[spaceElemId] = [elem]
+        }
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:50]: 😮 What is this magic?

[the remark this answers is not in the document]
Comment (Nurdok) [reply packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:50]: It's called [Nullish coalescing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment).

[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts:50]: Actually it works only in two lines
```        record[spaceElemId] ??= []
        record[spaceElemId].push(elem) ```

[the remark this answers is not in the document]
Comment (Nurdok) [reply packages/confluence-adapter/src/change_validators/unique_page_title_under_space.ts]: Weird, [I tried it in TS playground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBATgU1HAJgLhgJSSVAeaOASzAHMAaGMAVwFsAjBOAbQF0A+GAXhgG8BfANwAoABSJkKZgHIAhtNYwA-Ep5sAlADoADtQgALUQEZ1I8TlQz5ilWtZbdB0QCZTw0JBAAbBJq8hSc0l1GBgAejC+ACJZKIxmI0pnVn4gA) and it seems to work `¯\_(ツ)_/¯`

**Pull request merged**